File: Binary_Tree_Preorder_Traversal/binary_tree_preorder_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        if (root == null) return res;
        res.add(root.val);
        List<Integer> left = preorderTraversal(root.left);
        List<Integer> right = preorderTraversal(root.right);
        res.addAll(left);
        res.addAll(right);
        return res;
    }
}

File: Binary_Tree_Preorder_Traversal/binary_tree_preorder_traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

int countNodes(struct TreeNode* root) {
    if (root == NULL) {
        return 0;
    }
    return 1 + countNodes(root->left) + countNodes(root->right);
}

void traverse(struct TreeNode* node, int* res, int* index) {
    if (node == NULL) {
        return;
    }

    *(res + *(index)) = node->val;
    (*index)++;
    traverse(node->left, res, index);
    traverse(node->right, res, index);
}

int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    *returnSize = countNodes(root);

    int* res = malloc(sizeof(int) * countNodes(root));

     if (root == NULL) {
        return res;
    }

    int currentPosition = 0;
    traverse(root, res, &currentPosition);
    return res;
}


